### PR TITLE
Experimental Traffic Portal Self-Service UI

### PR DESCRIPTION
**Pull request by ocket8888**
## What does this PR do?
Introduces an Angular7-based subset of Traffic Portal, intended for use by users who want to monitor/configure limited parts of a syndicated CDN architecture. This is meant to first and foremost replace a certain, closed-source UI for which there is currently no open-source equivalent.

This project is overall a WIP, but I'd like it to be merged into `experimental/` in case anyone else wants to help out with the development.

Here's what's implemented so far:

- (mostly) mobile-friendly interface
- preliminary HTTPS support (only listens on localhost atm)
- login works and is persistent across page loads for as long as the `mojolicious` cookie is valid.
- unauthenticated users are directed to login
- API proxy fully implemented - any request that falls under `/api/` is sent upstream to TO
- Dashboard loads with a sticky nav bar and lists Delivery Service "cards" as details disclosure.
    - They aren't sorted, but can be quickly filtered by `displayName` using a fuzzy search bar
    - By default, the "cards" display only a summary consisting of the Display Name and Type, with some styling to indicate when a Delivery Service is not active (gets "greyed out" and appends `(inactive)` to the name), and an informational link that appears when the Delivery Service has an `infoURL` which, when clicked, opens said URL in a new browsing context (usually a "tab").
    - When the disclosures are opened - by clicking on them - further information is displayed, at the time of this writing consisting of:
         - The Delivery Service's "key" in the format `Key: {{xml_id}} ({{numeric id}})`
         - The `longDesc` field, if defined
         - A graph of edge bandwidth in KBPS since 00:00 UTC on the current day (refreshes whenever the disclosure is re-opened)
         - Meters measuring capacity (used/available/maintenance) and Cache Group Health. These display as `<meter>`s which generally render as a bar that goes from red to green as conditions improve, and include hover text and fallback rendering that display the actual percent as a number.
- The dashboard has a little FAB that you can use to create a new Delivery Service with limited configuration.
- "Users" page just lists all the users, containing the following information:
    - full name
    - username (currently used in the 'fuzzy search')
    - role
    - tenant
    - address (including street, city, state, zip)
    - contact info (including email/telephone)

## Which TC components are affected by this PR?

- [x] Introduces a new component: Traffic Portal Self-Service/Traffic Portal Angular7

## What is the best way to verify this PR?
A README with instructions for setup, deployment and development can be found at the project root (i.e. in [`experimental/traffic-portal-angular-7`](https://github.com/ocket8888/trafficcontrol/tree/angular-7/experimental/traffic-portal-angular7))

## Check all that apply

- [x] This PR includes tests (MUCH work to be done on them, however)
- [x] This PR includes all required license headers

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3357/
Test PASSed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3358/
Test PASSed.

**Comment by rawlinp** [comment]
Why do we need 2 UIs? Why can't we just have one UI with certain features hidden/turned off in order to make it "self-service"?

**Comment by ocket8888** [comment]
That's exactly the goal. Traffic Portal is currently kind of a mess, it's using a deprecated version of Angular (and several other things), depends on an external Sass compiler which has given us no end of trouble, and is by and large written using uncommented Javascript, which as you probably know is a bit of a nightmare to figure out. 

This is using an LTS version of Angular that compiles from easier-to-understand Typescript with type safety checks and mandatory JSDoc comments, has a built-in Sass compiler, and does server-side rendering. Truth be known I really just wanted to port Traffic Portal to this better framework, but it's believed that the project has a better chance of not being shot down immediately if it starts with an emphasis on Self-Service since that's something we want that the original Traffic Portal doesn't already do.

tl;dr: The idea is for this to eat Traffic Portal, so we **do** wind up with exactly one UI.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3360/
Test PASSed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3361/
Test PASSed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3362/
Test PASSed.

**Comment by mitchell852** [comment]
@ocket8888  threw this together as a lab week project. At first, he wanted to migrate TP from angular 1.6 to angular 7 but there is no migration path and that would be a large effort/large rewrite. Plus, TP was not built with self-service in mind. It's just a bunch of tables and forms. Some of the forms (i.e. the monsterous DS form) will never work for self-service imo. It's just way too complicated.

Therefore, I thought it would make more sense if he kept his scope smaller. I.e. build a UI (in angular 7), that focuses on the things required for self-service (i.e. ability to manage delivery services, users and tenants) and rethink how those things are managed. For example, maybe instead of simple (complex) forms, a wizard of some type makes more sense to simplify delivery services creation/update. I'm not sure. The point is if we start smaller with self-service in mind, we may end up with a very different UI than TP and then eventually we can port all the features from TP to this new UI......or maybe TP lives on as more of an admin tool. not sure.

**Comment by rawlinp** [comment]
Ok, I just wanted to make sure that we weren't going to continue on with two UIs again. We just managed to get rid of the old one :). My vote would be to fully replace TP with this once it gets feature parity, since when developing new features it's too much to ask to implement/test the feature in two different UIs. People are _still_ learning how to use the new TP UI. We will have to carefully consider how to get this TP v3 to work for both admin and self-service users.

**Comment by ocket8888** [comment]
> _"My vote would be to fully replace TP with this once it gets feature parity, since when developing new features it's too much to ask to implement/test the feature in two different UIs."_

That is my vote as well

> _"We will have to carefully consider how to get this TP v3 to work for both admin and self-service users."_ 

Yeah, I'd like to talk to some psuedo-self-service users, and actual admins (myself being only a dev) to try to figure out what they want out of a UI. I don't want to re-implement something they hate.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3367/
Test FAILed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3368/
Test FAILed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3371/
Test FAILed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3372/
Test FAILed.

**Comment by ocket8888** [comment]
Tests are failing because of #3421

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3375/
Test FAILed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3400/
Test FAILed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3404/
Test FAILed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3406/
Test FAILed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3407/
Test FAILed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3415/
Test PASSed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3417/
Test PASSed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3422/
Test PASSed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3437/
Test PASSed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3438/
Test PASSed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3439/
Test PASSed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3440/
Test PASSed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3441/
Test PASSed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3449/
Test FAILed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3450/
Test PASSed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3451/
Test PASSed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3452/
Test PASSed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3453/
Test PASSed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3455/
Test PASSed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3474/
Test PASSed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3486/
Test PASSed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3519/
Test FAILed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/3528/
Test PASSed.

**Comment by mitchell852** [comment]
Brennan did the following with this PR:

* Enabled server-side rendering

* Made a login page

* Started working on login; implemented basic API proxy

* Implemented login functionality

* Log-in now works, dashboard displays list of DS display Names

* Added typedoc documentation comments

* Fixed mixed export state merged declarations

* Login page is now pretty

* Fixed login form validation

* Started Delivery Service cards

* Added DS cards to dashboard, improved header layout, introducing sample graphs!

* Login state now persistent across pageloads (while mojo cookie is valid)

* Added alerts skeleton; added capacity meters to ds cards

* Added bandwidth graphs to DS cards

* Added instructions for running in debug mode, and added requisite proxy definiton file to ignore

* Added tabindex/accesskey to search input, changed dummy nav links to things that will actually get used in the future

* Added a dirty hack to make the loading cursor display over the ds-card area while loading.

* Fixed an issue where graphs won't display against old TO versions

Also added graph titles, fixed protocols not being displayed as strings,
added a proper loading area when chart data is loading, fixed some ds-card
positioning and relative spacing, and added default text when a DS doesn't
have a description ('NO DESCRIPTION').

* Moved TP Angular 7 into experimental

* Added contribution guidelines (and made code-base compliant with said guidelines)

* Added browser support mission statement.

* Added fix for Blink render engine flex/grid implementation

* Fixed getDSKBPS not respecting 'interval' argument

* Added shims for browsers that may not support min/max content sizing

* Added secondary and tertiary descriptions to ds-cards

* Search bar now sticks inside header

* Fixed login not working in debug mode

* Fixed autonomous builds outputting unprintable characters

* Added skeleton '/users' page

* Added missing license header

* Added default warning styling

* Fixed improper modeling of Delivery Services' 'exampleURLs' field

* Fixed poor handling of missing data

* Added URL listing to ds cards

* Better handling of no-cachegroup-having delivery services

* Fully implemented visual alerts

* Fixed all DS-cards sharing a graph

* Fixed DS-cards redrawing over their own graphs without erasing first

* Fixed alt text overflowing container in the header element

* Started working on a new Delivery Service form

* Rolled API version back to 1.4; addeda method for getting types

* Buttons now use pointer cursor by default (disabled cursor when disabled)

* Added CDN model and added CDN fetch method to API

* Added missing 'rangeRequestHandling' field and field formatting

* DOM access by alert component is now truly platform-independent

* Added accessibility title to new DS button

* Fixed DSes with no graph data always having loading cursors

* Finished basic 'new DS' form functionality

* Added 'previous' buttons and finished 'new DS' form styling

* Fuzzy Search now case-insensitive

* Implemented user fuzzy search

* Componetized header

* Added a static logo asset

I was frustrated by the viewbox setting on our 'misc' one - plus, this
is the only way to get it to show up in debug mode/dev env.

* Added and implemented loading animation

* Fixed submission of a new Delivery Service not redirecting properly

* Added model for Roles (i.e. 'Role models', heh)

* Added API method for fetching roles from TO

* Added a user card component

* Implemented user cards on the '/users' page

* Removed some debug logging and improved linting compliance

* Fixed graphs overlaying on top of fuzzy search bar

* Fixed header overlaying alerts

* Removed heinous debug logging

* fixed fuzzy search on Users page not properly hiding users

* Visual improvement to user cards

* Fixed user cards taking forever to fetch role names

Previously, each card made its own HTTP request to get a name for its Role.
The new implementation makes only 1 request at the users page level to
fetch a map of role IDs to role names. This behaviour is optional when using
a user-card, and it will fall back to the old behaviour if no map is passed.

* Added generated documentation directory to gitignore

* Now 100% more configurable at run time!

The server executable now offers options to change the port on which
it listens, the TO server information, and certificate verification
(as well as standard version and help options). Also, the server
will now check connectivity with the specified TO server at startup
(via `/api/1.4/ping`), and will exit with an error if the check fails.

* enabled preliminary ssl support

* Added the ability to fetch delivery service types (will also cache)

* Added advanced option for changing DS type

* Mobile-friendly update

Navbar now collapses on small screens, fixed some overlay issues
with ds charts and the header and title text etc.

* Fixed presentation on very small mobile screens

* Improved test coverage

* Added testing support for Firefox and Opera

* Docs fixup for testing

* Removed unused input

* minor testing improvements

* Added some customization options to the 'New Delivery Service' form

* Fixed code scaffolding to obey linting rules

* Added some advanced fields to the new DS wizard

* Added a 'view-details' button to ds-cards
